### PR TITLE
Added UDF for emotion detection

### PR DESCRIPTION
**Pull request by Anirudh58**

### CODE REVIEW

**Review by gaurav274 — CHANGES_REQUESTED** [review]

File: eva/udfs/emotion_detector.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+        # load model
+        self.model = VGG("VGG19")
+        model_state = torch.load(
+            os.path.join(EVA_DEFAULT_DIR, "data", "models", "emotion_detector.t7")
+        )
+        self.model.load_state_dict(model_state["net"])
+
+        # move to GPU and set to evaluation mode
+        self.to_device("0")
```
[the remark this answers is not in the document]
Comment (Anirudh58) [reply eva/udfs/emotion_detector.py:136]: Hmm without this, I get an error saying input tensors are CUDA but the model is not. I am inheriting `AbstractClassifierUDF` and implementing the `classify` method.

File: eva/udfs/emotion_detector.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        return "EmotionDetector"
+
+    def __init__(self, threshold=0.85):
+        super().__init__()
+        self.threshold = threshold
+
+        # load model
+        self.model = VGG("VGG19")
+        model_state = torch.load(
+            os.path.join(EVA_DEFAULT_DIR, "data", "models", "emotion_detector.t7")
```
[the remark this answers is not in the document]
Comment (Anirudh58) [reply eva/udfs/emotion_detector.py:131]: Oh yes sure, I can do that. But what should the path here be then?

**Comment by Anirudh58** [comment]
> @Anirudh58 I shared a link to the Google drive folder. We should also add some test cases.

Right I'll upload the model to the drive. And yes I'll add a few testcases for this UDF

File: eva/udfs/emotion_detector.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gaurav274) [reply eva/udfs/emotion_detector.py:136]: `def to_device(self, device: str):` should have moved the model to GPU. Can you verify if it is called?

[the remark this answers is not in the document]
Comment (Anirudh58) [reply eva/udfs/emotion_detector.py:136]: Yes I tried that. It does not automatically move. `to_device()` is not being called.

File: eva/udfs/emotion_detector.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+import torchvision.transforms as transforms
+from PIL import Image
+from skimage.transform import resize
+from torch.autograd import Variable
+
+from eva.configuration.dictionary import EVA_DEFAULT_DIR
+from eva.udfs.abstract_udf import AbstractClassifierUDF
+from eva.udfs.gpu_compatible import GPUCompatible
+
+# VGG configuration
```
Comment (LordDarkula) [inline eva/udfs/emotion_detector.py:33]: What kind of configuration is this? Is this the structure of the neural network? A small comment would be nice

**Review by LordDarkula — CHANGES_REQUESTED** [review]

File: eva/udfs/emotion_detector.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        512,
+        512,
+        512,
+        "M",
+    ],
+}
+
+
+# helper class for VGG
+class VGG(nn.Module):
```
[the remark this answers is not in the document]
Comment (gaurav274) [reply eva/udfs/emotion_detector.py:84]: This should be fine as I guess

File: eva/udfs/emotion_detector.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                    nn.Conv2d(in_channels, x, kernel_size=3, padding=1),
+                    nn.BatchNorm2d(x),
+                    nn.ReLU(inplace=True),
+                ]
+                in_channels = x
+        layers += [nn.AvgPool2d(kernel_size=1, stride=1)]
+        return nn.Sequential(*layers)
+
+
+class EmotionDetector(AbstractClassifierUDF, GPUCompatible):
```
Comment (gaurav274) [inline eva/udfs/emotion_detector.py:114]: We can maybe inherit this from `PytorchAbstractClassifierUDF`, and then `to_device` should be taken care of automatically.

**Comment by gaurav274** [comment]
@Anirudh58 @LordDarkula We need to close this PR asap. It is important for the release.

File: eva/udfs/emotion_detector.py (hunk 3 above)
Comment (Anirudh58) [reply eva/udfs/emotion_detector.py]: Yes, it specifies the architecture. But yes, let me clean it up a little.

**Comment by jarulraj** [comment]
We are planning to use this approach for downloading models using Dropbox links -- 
https://github.com/Emilien-mipt/fer-pytorch/blob/main/fer_pytorch/pre_trained_models.py#L12

**Comment by gaurav274** [comment]
@Anirudh58 Are we planning to make the changes Joy mentioned in this PR?

**Comment by jarulraj** [comment]
> @Anirudh58 Are we planning to make the changes Joy mentioned in this PR? We can also pass the `model_dir` in the `torch.utils.model_zoo.load_url` function. How about we keep all these models in the `~/.eva/udfs/models/`? @LordDarkula @jarulraj

Sure, that should work.

**Review by gaurav274 — APPROVED** [review]

**Pull request merged**